### PR TITLE
feat(debugger): make element explorer work with node 0.12.0

### PR DESCRIPTION
**Pull request by hankduan**
Node has changed its debugger significantly in 0.12.0, and these changes are necessary to get it to work now.

Specifically here are the key points to take note:

* Before, the process continues running after `process._debugProcess(pid);` is called, but now, the process stops. 
To over come this, the call to `process._debugProcess(pid)` is moved from protractor into the debugger client. Secondly, because the process stops after the call, we call `reqContinue` once the debugger connection is established, so that protractor continues into the first break point (for backwards compatibility, we added an extra empty webdriver command so that in earlier versions of node, protractor doesn't go past the first break point from the reqContinue). 

* Before repl provides '(foobar\n)' when an user inputs 'foobar'. Now it is just 'foobar\n'.
We will parse and strip away both the parenthesis and '\n' to support all versions of node. 

Additionally, this change makes debugger processes fail fast if the port is taken.

### CODE REVIEW

File: lib/debugger/debuggerCommons.js [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+  client.once('ready', function() {
+    client.setBreakpoint({
+      type: 'scriptRegExp',
+      target: '.*executors\.js', //jshint ignore:line
+      line: 37
+    }, function() {});
+  });
+
+  process.debugPort = opt_port || 5858
```
Comment (tullmann) [inline lib/debugger/debuggerCommons.js]: Missing a trailing ';' here.

File: lib/debugger/debuggerCommons.js [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  var connectWithRetry = function(attempts) {
+    var socket = client.connect(process.debugPort, host);
+    socket.on('connect', function() {
+      client.reqContinue(function() {
+        // Intentionally blank.
+      });
+    })
+    socket.on('error', function(e) {
+      if (attempts === 1) {
+        throw e
```
Comment (tullmann) [inline lib/debugger/debuggerCommons.js]: Missing a trailing ';' here.

**Comment by hankduan** [comment]
oops, forgot to run through linter. Made some style changes thanks @tullmann

**Comment by floribon** [comment]
That indeed unlocked the `elementExplorer` (and `browser.pause()`) for me.
However, commands take a while and freeze most of the time. I'm using iojs@1.4.3 which forked node 0.10.35

**Comment by hankduan** [comment]
Node versions above 0.10.30 is affected by the nodejs bug described here: https://github.com/angular/protractor/issues/1890

**Comment by floribon** [comment]
Yes but I was experiencing a different bug, the debugger wouldn't work at all, it was impossible to enter any command. With these commits I fall back to that sluggish bug indeed. Using protractor 1.6.1 didn't work for me either, so just saying for any potential iojs user. Thanks for your work.

**Comment by hankduan** [comment]
@juliemr when you review this, please take note of how I check the port for whether it is in use or not, as I'm not sure if that's the best way to do it.

**Comment by sibeliusseraphini** [comment]
+1

File: lib/debugger/clients/explorer.js [hunk 3]
```diff
@@ -9,7 +9,7 @@ var CommandRepl = require('../modes/commandRepl');
  * @constructor
  */
 var WdRepl = function() {
-  this.client = new baseDebugger.Client();
+  this.client;
```
Comment (sjelin) [inline lib/debugger/clients/explorer.js]: This line appears to do nothing and could be removed

**Comment by bootstraponline** [comment]
:+1:

File: lib/debugger/clients/wddebugger.js [hunk 4]
```diff
@@ -103,6 +99,11 @@ WdDebugger.prototype.initRepl_ = function() {
 
   // We want the prompt to show up only after the controlflow text prints.
   this.dbgRepl.printControlFlow_(function() {
+    // Backward compatibility: node version 0.8.14 has a number of built in
```
Comment (juliemr) [inline lib/debugger/clients/wddebugger.js]: I don't think we need to worry about versions <10 anymore, feel free to remove this.

File: lib/protractor.js [hunk 5]
```diff
@@ -661,34 +661,30 @@ Protractor.prototype.initDebugger_ = function(debuggerClientPath, opt_debugPort)
...
-    if (self.debuggerServerPort_) {
-      args.push(self.debuggerServerPort_);
-    }
-    var nodedebug = require('child_process').fork(debuggerClientPath, args);
-    process.on('exit', function() {
-      nodedebug.kill('SIGTERM');
+  // Invoke fn if port is available.
+  var onPortAvailable = function(port, fn) {
+    var net = require('net');
+    var tester = net.connect({port: port}, function() {
```
Comment (juliemr) [inline lib/protractor.js]: Can you check that the error is EADDRINUSE?

File: lib/protractor.js [hunk 6]
```diff
@@ -661,34 +661,30 @@ Protractor.prototype.initDebugger_ = function(debuggerClientPath, opt_debugPort)
...
 
   var vm_ = require('vm');
+  var flow = webdriver.promise.controlFlow();
+
   var context = { require: require };
-  for (var key in global) {
-    context[key] = global[key];
-  }
-  context.list = function(locator) {
+  global.list = function(locator) {
```
Comment (juliemr) [inline lib/protractor.js]: Why this change? I think it's safer to avoid polluting `global.list`

Comment (hankduan) [inline lib/protractor.js]: otherwise autocomplete doesn't suggest it, since autocomplete doesn't know about context.

Comment (juliemr) [inline lib/protractor.js]: I see. I don't think we need autocomplete to suggest it, since it's all of 4 characters, but I don't really care either way.

File: lib/protractor.js [hunk 7]
```diff
@@ -700,9 +696,33 @@ Protractor.prototype.initDebugger_ = function(debuggerClientPath, opt_debugPort)
...
+      process.on('exit', function() {
+        nodedebug.kill('SIGTERM');
+      }); 
+    });
+  });
+  
+  var pausePromise = flow.timeout(1000, 'waiting for debugger to attach')
+      .then(function() {
+        // Necessary for backward compatibility with node < 0.12.0
+        browserUnderDebug.executeScript_('', 'empty debugger hook');
```
Comment (juliemr) [inline lib/protractor.js]: Why wasn't this necessary before?

File: lib/debugger/clients/wddebugger.js (hunk 4 above)
Comment (hankduan) [inline lib/debugger/clients/wddebugger.js]: Keeping this, as discussed offline.

**Comment by juliemr** [comment]
Looks good, just a couple questions to make sure I understand the changes.

File: lib/protractor.js (hunk 5 above)
Comment (hankduan) [inline lib/protractor.js]: This is not an error. I'm testing that I _can_ connect to it and fail if so. 

I tried something like https://gist.github.com/timoxley/1689041 before, where I'm test creating a server to check for EADDRINUSE. However, that does not fail for some reason, until I try to create the debug process.

Comment (juliemr) [inline lib/protractor.js]: Ah sorry, I meant something like this: https://github.com/SeleniumHQ/selenium/blob/master/javascript/node/selenium-webdriver/net/portprober.js#L159

Except we don't actually want to create a server, so the method you had is the best I can think of.

File: lib/protractor.js (hunk 7 above)
Comment (hankduan) [inline lib/protractor.js]: It's because the way node debugger changed in 0.12.
Before, when you start the debug process, it starts a server but the program continues (that's why you needed the sleep() in there to manually stop execution while waiting for the debugger to attach). However, now, the moment you start debug, the program stops and waits for the reqContinue signal. 

As a result of that, before, we call debug in this process. In the client, we take control once we attach.
Now, we call debug from the client. Because the program is paused, reqContinue is the first call we make. After that, we can take control as before. 

However, if we're using node 0.10, the program isn't paused when issued the debug, so the first reqContinue actually makes the program jump to the next break point.

**Comment by hankduan** [comment]
merged: https://github.com/angular/protractor/commit/de499696eb88721cb073a195025ae48f59cbc905
